joern-cli/frontends/javasrc2cpg/src/main/scala/io/joern/javasrc2cpg/util/Scope.scala

```diff
@@ -1,46 +1,125 @@
 package io.joern.javasrc2cpg.util
 
-import io.joern.javasrc2cpg.util.Scope.WildcardImportName
-import io.joern.x2cpg.datastructures.Stack.Stack
-import io.joern.x2cpg.datastructures.{ScopeElement, Stack, Scope => X2CpgScope}
-import io.shiftleft.codepropertygraph.generated.nodes.{HasTypeFullName, NewNode, NewTypeDecl}
+import io.joern.javasrc2cpg.passes.ExpectedType
+import io.joern.javasrc2cpg.util.Scope.ScopeTypes.{MethodScope, NamespaceScope, ScopeType, TypeDeclScope}
+import io.joern.javasrc2cpg.util.Scope.{VariableNodeType, WildcardImportName}
+import io.joern.x2cpg.Ast
+import io.joern.x2cpg.datastructures.{ScopeElement, Scope => X2CpgScope}
+import io.shiftleft.codepropertygraph.generated.nodes.{
+  DeclarationNew,
+  HasName,
+  HasNameMutable,
+  HasTypeFullName,
+  NewLocal,
+  NewMember,
+  NewMethod,
+  NewMethodParameterIn,
+  NewNamespaceBlock,
+  NewNode,
+  NewTypeDecl
+}
+import org.slf4j.LoggerFactory
 
 import scala.collection.mutable
 
-case class NodeTypeInfo(node: NewNode with HasTypeFullName, isField: Boolean = false, isStatic: Boolean = false)
+case class NodeTypeInfo(node: VariableNodeType, isField: Boolean = false, isStatic: Boolean = false)
 
-class Scope extends X2CpgScope[String, NodeTypeInfo, NewNode] {
+class Scope extends X2CpgScope[String, NodeTypeInfo, ScopeType] {
 
-  private var typeDeclStack: List[NewTypeDecl] = Nil
+  private val logger                                        = LoggerFactory.getLogger(this.getClass)
+  private var typeDeclStack: List[NewTypeDecl]              = Nil
+  private var lambdaMethods: List[mutable.ArrayBuffer[Ast]] = Nil
+  private var lambdaDecls: List[mutable.ArrayBuffer[Ast]]   = Nil
 
-  override def pushNewScope(scopeNode: NewNode): Unit = {
-    scopeNode match {
-      case typeDecl: NewTypeDecl => typeDeclStack = typeDecl :: typeDeclStack
-      case _                     => // Nothing to do in this case
+  override def pushNewScope(scope: ScopeType): Unit = {
+    scope match {
+      case TypeDeclScope(declNode) =>
+        typeDeclStack = declNode :: typeDeclStack
+        lambdaMethods = mutable.ArrayBuffer[Ast]() :: lambdaMethods
+      case NamespaceScope =>
+        lambdaDecls = mutable.ArrayBuffer[Ast]() :: lambdaMethods
+      case _ => // Nothing to do in this case
     }
 
-    super.pushNewScope(scopeNode)
+    super.pushNewScope(scope)
+  }
+
+  def getCapturedVariables: List[NewNode with HasTypeFullName with HasNameMutable] = {
+    stack
+      .foldLeft(List.empty[NewNode]) { (acc, stackEntry) =>
+        acc ++ stackEntry.variables.values.map(_.node)
+      }
+      .collect {
+        case param: NewMethodParameterIn => param
+        case local: NewLocal             => local
```

Comment:
if i'm not mistaken, a _case_ for `NewMember` is missing here

Reply:
I originally had one there, but don't think there should be. The justification is that, for lambdas using object members, a `this` reference is added as the 0th lambda param (consistent with the java compiler), so the member would be accessed as `this.member`.